BUG: astype falsely converts inf to integer, patch for Numpy (GH14265)

shawnheide opened the pull request:
- [x] closes #14265
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

Bug in Numpy causes inf values to be falsely converted to integers. I added a ValueError exception similar to the exception for trying to convert NaN to an integer.

jorisvandenbossche commented on pandas/types/cast.py:
```diff
@@ -529,6 +529,10 @@ def _astype_nansafe(arr, dtype, copy=True):
 
         if np.isnan(arr).any():
             raise ValueError('Cannot convert NA to integer')
+        elif np.isinf(arr).any():
```
I think we can combine both checks using one `not np.isfinite(arr).all()`

shawnheide committed:
BUG: astype falsely converts inf to integer, patch for Numpy (GH14265) (+2 squashed commit)
Squashed commits:
Update test messages to catch new ValueError
Change sparse test for astype to assertRaises(ValueError) instead of regex version

jorisvandenbossche commented on pandas/types/cast.py:
```diff
@@ -527,8 +527,9 @@ def _astype_nansafe(arr, dtype, copy=True):
-        if np.isnan(arr).any():
-            raise ValueError('Cannot convert NA to integer')
+        if not np.isfinite(arr).all():
+            raise ValueError('Cannot convert non-finite values to integer')
```
can you add between brackets after "non-finite" something like "(NA or inf)", then it is clearer what is meant

shawnheide replied:
Yeah, no problem. I debated adding it, but went for succinctness. I agree it is informative, adding it now.

jorisvandenbossche commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -79,3 +79,5 @@ Performance Improvements
 Bug Fixes
 ~~~~~~~~~
+
+- Bug in Numpy ``astype()`` caused `inf` values to be falsely converted to integers. Affected ``astype()`` in Series and DataFrames (:issue:`14265`)
```
I wouldn't call this a "bug in numpy". It is defined behaviour of numpy, which we just don't want to expose to our users and therefore provide an error (as we already do with NaN)

jorisvandenbossche reviewed (commented):
Two small comments, looks good for the rest!

jorisvandenbossche commented:
Looks good to me 
@jreback

jorisvandenbossche commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -79,3 +79,5 @@ Performance Improvements
 Bug Fixes
 ~~~~~~~~~
+
+- Numpy ``astype()`` caused `inf` values to be falsely converted to integers. Affected ``astype()`` in Series and DataFrames (:issue:`14265`)
```
@shawnheide Can you add that it now raises an error instead?

jreback replied:
use double-backticks on ``inf``

jreback replied:
say something more like:

Bug in ``.astype()`` where ``inf`` values incorrectly converted to integers.

jreback commented on pandas/tests/types/test_cast.py:
```diff
@@ -203,6 +203,18 @@ def test_possibly_convert_objects_copy(self):
+    def test_astype_raises(self):
+        types = [np.int32, np.int64]
+        values = [np.nan, np.inf]
+
```
this needs to go in tests/series/tests_dtypes.py

jorisvandenbossche replied:
Aha, I already found it strange I couldn't see any other astype tests

jreback commented on pandas/tests/types/test_cast.py:
```diff
@@ -203,6 +203,18 @@ def test_possibly_convert_objects_copy(self):
+            for this_val in values:
+                s = Series([this_val])
+                self.assertRaises(ValueError, s.astype, this_type)
+
```
add a test on DataFrames in the analagous file in frames.

shawnheide commented:
Sorry @jorisvandenbossche and @jreback it took so long to make these simple fixes, but I just got them done. I ended up merging the tests in series and frame with the previous test for NaN conversion. Let me know if you have any more requests. Cheers.

jreback commented:
can you rebase (so can check with current code)

shawnheide commented:
Just rebased and re-ran the tests, everything looks good. Let me know if there's anything else I need to do.

jreback commented on pandas/sparse/tests/test_array.py:
```diff
@@ -361,8 +361,7 @@ def test_astype(self):
 
         arr = SparseArray([0, np.nan, 0, 1], fill_value=0)
-        msg = "Cannot convert NA to integer"
-        with tm.assertRaisesRegexp(ValueError, msg):
```
isn't the message the same?

jreback replied:
if u want to assert the new message would be great

shawnheide commented:
Thanks for catching that @jreback. Hopefully that should do it.

jorisvandenbossche commented:
@shawnheide Thanks!
